libnvme/src/firmware.rs

```diff
@@ -0,0 +1,351 @@
+    ControllerError(#[from] NvmeControllerError),
+    #[error(
+        "libnvme says the log page is {} bytes but it should be {} bytes",
+        size,
+        expected_size
+    )]
+    UnexpectedSize { size: usize, expected_size: usize },
+}
+
+impl<'a> Controller<'a> {
+    /// Get the controller's firmware log page.
+    pub fn get_firmware_log_page(
+        &self,
+    ) -> Result<FirmwareLogPage, FirmwareLogPageError> {
+        let expected_size = mem::size_of::<nvme_fwslot_log_t>();
+        let LogPageInfo { size, req, .. } =
+            self.get_logpage(LogPageName::Firmware)?;
+        if size != expected_size {
+            return Err(FirmwareLogPageError::UnexpectedSize {
+                size,
+                expected_size,
+            });
+        }
+        let mut buf = MaybeUninit::<nvme_fwslot_log_t>::zeroed();
+        self.check_result(
+            unsafe {
+                nvme_log_req_set_output(
+                    req.inner,
+                    buf.as_mut_ptr().cast(),
+                    size,
+                )
+            },
+            || format!("failed to set logpage req size to {size}"),
+        )?;
+
+        self.check_result(unsafe { nvme_log_req_exec(req.inner) }, || {
+            "failed to execute firmware log request"
+        })?;
+
+        // Our request was successful so we can assume this struct has been
+        // populated now.
+        let logpage = unsafe { buf.assume_init() };
+        let controller_info = self.get_info()?;
+        let identify = controller_info.get_controller_info_identify();
+        Ok(FirmwareLogPage::init(&identify, logpage))
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum FirmwareLoadError {
+    #[error("{0}")]
+    Nvme(#[from] NvmeError),
+    #[error("{0}")]
+    NvmeController(#[from] NvmeControllerError),
+    #[error("Failed to upload firmware: {0}")]
+    IoError(#[from] std::io::Error),
+    #[error("Supplied firmware is too large")]
+    FirmwareImageTooLarge,
+}
+
+impl<'ctrl> WriteLockedController<'ctrl> {
+    /// Load a single chunk of firmware at offset
+    fn firmware_load_chunk(
+        &self,
+        data: &[u8],
+        offset: u64,
+    ) -> Result<(), NvmeControllerError> {
+        self.check_result(
+            unsafe {
+                nvme_fw_load(
+                    self.inner,
+                    data.as_ptr().cast(),
+                    data.len(),
+                    offset,
+                )
+            },
+            || "failed to load firmware",
+        )
+    }
+
+    /// Upload new firmware to the NVMe controller.
+    ///
+    /// Note this firmware needs to be commited to a slot via
+    /// `FirmwareCommitRequestBuilder`.
+    pub fn firmware_load<R: Read>(
+        &self,
+        mut data: R,
+    ) -> Result<(), FirmwareLoadError> {
+        // TODO swap to the libnvme granularity function Andy is adding
+        const CHUNK_SIZE: u64 = 0x1000;
+        let mut offset = 0u64;
+
+        loop {
```

Comment:
Thanks for mentioning the nuance of `read_exact` leaving the buffer in an unspecified state.

I have opted to read the entire thing into a Vec and split it on chunks.  While this does mean we are reading things into memory it makes chunking the bytes easier. I will have a follow up commit to address #7 

Additionally I think we should possibly add a `take` call on the reader before the call to `read_to_end` so that we can stop a library consumer from passing us something huge. I need to look in the spec to see if there is anything about a maximum size otherwise we can use a conservative value that should work.